### ISSUE TITLE
Adding weights to loss function in case of imbalaced dataset

### ISSUE THREAD

**Issue by asdspal** [pytorch/ignite#361]
How can I add weight to the loss function in case of imbalanced dataset?

**Comment by asdspal** [comment]
Thanks for replying. Normally, I would define my weighted loss as:
`weights =  torch.tensor([1,10])`
`loss = torch.nn.BCEWithLogitsLoss(true, output, weight=weights[true])`

The size of the weight should be equal to batch_size. I need the true labels for the weights.  
How do I do that?

**Comment by vfdev-5** [comment]
@asdspal there are several ways achieve this:
1) If you use `create_supervised*` methods to construct a trainer, idea is to compute the sample weights in dataloader
```python
# define a dataloader that returns a batch as (x, y, weights)
dataloader = ...

def loss_fn(y_pred, y):
    y_true, weights = y
    return torch.nn.BCEWithLogitsLoss(y_pred, y_true, weights=weights)

trainer =  create_supervised_trainer(model, optimizer, loss_fn)
```
2) Second method may be more more verbose but allows to fully control model updating step
```python
from ignite.engine import Engine, _prepare_batch

loss_fn = torch.nn.BCEWithLogitsLoss

def update(engine, batch):
    model.train()
    optimizer.zero_grad()
    x, y = _prepare_batch(batch, device='cuda', non_blocking=True)
    y_pred = model(x)
    weights = torch.tensor([1,10])
    loss = loss_fn(y_pred, y, weight=weights[true])
    loss.backward()
    optimizer.step()
    return loss.item()

trainer = Engine(update)
```

HTH

**Comment by asdspal** [comment]
Many thanks for clearing this up for me. I have gone with approach one and it's working.

**Comment by vfdev-5** [comment]
Okay, I'm closing this issue as problem is solved, feel free to reopen it if you have more questions on this problem.
